lib/Dialect/FIRRTL/Export/FIREmitter.cpp

```diff
@@ -1087,6 +1134,20 @@ void Emitter::emitStatement(InvalidValueOp op) {
   emitLocationAndNewLine(op);
 }
 
+void Emitter::emitStatement(GenericIntrinsicOp op) {
+  startStatement();
+  if (op.use_empty())
+    emitGenericIntrinsic(op);
+  else {
```

Comment:
Could we also emit the intrinsic inline if it has a single use?

Reply:
See `isEmittedInline` -- and the answer is "nearly" (I had it implemented that way at first) -- but only if we know the use will be emitted (exactly) once.

To address this, the idea is to only emit it inline if it's used exactly once and that use itself is not emitted inline.

Probably this should be pulled into a pre-pass that inserts declarations as needed and generates names for declarations that don't have any, so on.  That's left as future work and this just tries to get good output using a simple but correct heuristic.  But please look carefully :+1:.

Reply:
On the subject, this "inline" suitability is arguably (?) not quite right re:re-ordering operations that are side-effecting if the appear in one order but are used in a different order.  I don't think it's worth worrying about that as what they "effect" is outside the design's dataflow and state and is unreachable without some effort, but just a thought.  I think all our intrinsics are pure when expression-like (have a result, and possibly a use) presently.

Reply:
Oh yeah I hadn't considered that this might be emitted as part of a larger expression that potentially gets inlined multiple times. Better be conservative. And your point about inlining affecting the order of execution is also a good one. I know we don't really give any guarantees in FIRRTL, but no need to artificially move stuff around for no reason. Emission into a node looks great!